internal/consul/connection.go

```diff
@@ -11,40 +32,183 @@ type Client interface {
 	DiscoveryChain() *api.DiscoveryChain
 	Namespaces() *api.Namespaces
 
+	WatchServers(ctx context.Context) error
+
+	Token() string
+	Wait(until time.Duration) error
+
 	// TODO: drop this
 	Internal() *api.Client
 }
 
+type ClientConfig struct {
+	ApiClientConfig *api.Config
+	PlainText       bool
+	Addresses       string
+	HTTPPort        int
+	GRPCPort        int
+	TLS             *tls.Config
+	Credentials     discovery.Credentials
+	Logger          hclog.Logger
+}
+
 type client struct {
-	client *api.Client
+	stop        func()
+	config      ClientConfig
+	client      *api.Client
+	token       string
+	mutex       sync.RWMutex
+	initialized chan error
 }
 
-func NewClient(c *api.Client) Client {
+func NewClient(config ClientConfig) Client {
+	config.Logger = hclog.Default()
 	return &client{
-		client: c,
+		config:      config,
+		initialized: make(chan error, 1),
+	}
+}
+
+func (c *client) Wait(until time.Duration) error {
+	select {
+	case err := <-c.initialized:
+		return err
+	case <-time.After(until):
+		c.stop()
+		return errors.New("did not get state within time limit")
+	}
+}
+
+func (c *client) WatchServers(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	c.stop = cancel
+
+	var static bool
+	serverName := c.config.Addresses
+	if strings.Contains(serverName, "=") {
+		serverName = ""
+	} else {
+		static = true
+	}
+	if c.config.TLS != nil && c.config.TLS.ServerName != "" {
+		serverName = c.config.TLS.ServerName
+	}
+
+	config := discovery.Config{
+		Addresses:   c.config.Addresses,
+		GRPCPort:    c.config.GRPCPort,
+		Credentials: c.config.Credentials,
+	}
+
+	if !c.config.PlainText {
+		config.TLS = c.config.TLS
+	}
```

Comment:
I was curious about the best way to check for whether TLS should be used for these gRPC connections - this approach seems reasonable, in 
I took a different approach more similar to `internal/commands/server/server.go` and `internal/commands/exec/exec.go` though, checking
```
if consulAPITLSConfig.CAFile != "" || len(consulAPITLSConfig.CAPem) > 0
```
It looks like consul-dataplane is asking for an explicit `-tls-disabled` configuration https://github.com/hashicorp/consul-dataplane/#configuration-reference